### ISSUE TITLE
How to manage permissions and user roles

### ISSUE THREAD

**Issue by samuelgfeller** [odan/slim4-tutorial#56]
In #34 you told me that it'd make sense to do the authorization logic in the service class.  

Up until now, I kept user role management simple by having a string with the role name in the `user` table. It is clear that they should be put in an individual table `user_role`. I want to know your opinion of the following approaches that I consider implementing. 

### 1. Have specific authorization logic in each service class
This would be how I did it until now and would look like the following:
```php
final class ClientUpdater
{
   public function __construct(/* ... */) { }

   public function updateClient(int $clientIdToChange, array $clientValues, int $loggedInUserId): bool
    {
        $client= new ClientData($clientValues);
        $this->clientValidator->validateClientUpdate($clientIdToChange, $client);

        $userRole = $this->userRepository->getUserRoleById($loggedInUserId);
        // Check if authenticated user is admin or if he's linked to the client
        if ($userRole === UserRole::ADMIN || $client['user_id'] === $loggedInUserId) {
            // Update logic
            // return;
        }
        // User does not have needed rights to access area or function
        $this->logger->notice('User ' . $loggedInUserId . ' tried to update client with id: ' . $userIdToChange . ' but is not allowed.');
        throw new ForbiddenException('Not allowed to update client.');
    }
}
```

### 2. Store permission rules in database
When searching on the web, I found this solution that I kind of like by https://code-boxx.com/php-user-role-management-system/.  

<img src="https://github.com/samuelgfeller/slim-example-project/assets/31797204/9c39fddc-768a-4893-92e5-477dd495f60e" width="400">
 
This seems like a more classic approach that I could imagine is better for large scaled projects. Each action would be described in a table, and in another one is the link between which user roles can perform which actions. 

Testing is an important thing for me, it may be easier to test the features with different user roles if the authorization logic is clearly defined in the database and not in each specific's action service class. 

But I am not sure, and I have no experience with well-made large applications. That's why I'm asking you. I love the use case based approach and keeping each thing in the smallest "unit" that makes sense.

**Comment by odan** [comment]
There is a difference between authentication and authorization. **Authentication** is needed to check if a user is allowed to log in and logged-in. **Authorization** is used to check if a logged-in user has permissions to access a specific resource.
Making HTTP authentication check belongs to the Action class (or Controller layer in MVC). The thing is, depending on your implementation, you also need a database connection and some "logic" to check the user (or API) credentials. This is why I said I would put this auth logic into a specific class.

~~In your example, you are mixing authentication and authorization (RBAC)~~. To keep it more clean, I would move at least the authentication check (is the user logged in) into a middleware because this is an HTTP specific concern.

**Comment by samuelgfeller** [comment]
I'm aware of the difference, and I have the authentication in a middleware that automatically redirects to the login page if the request has no active session on a protected route. The service class `updateClient` cannot be reached without an active session.

Here, I'm specifically asking for which authorization model you would recommend.
I mean by that, the part where you check if an authenticated user is allowed to perform a specific action, like updating `client` data. An `admin` for example should be allowed to edit any data even if he is not the author. A `user` on the other hand can only edit the client if he is assigned to him by being linked via `user_id`. 

The question is this check, how do I do it? I have two quite different approaches that I consider implementing. The first is via `if` statements, checking if the user role is admin or if the user is the resource owner: 
```php
        $userRole = $this->userRepository->getUserRoleById($loggedInUserId);
        // Check if authenticated user is admin or if he's linked to the client
        if ($userRole === UserRole::ADMIN || $client->user_id === $loggedInUserId) {
            // Update logic
            // return;
        }
        throw new ForbiddenException('Not allowed to update client.');
```
The other one is illustrated and explained in this article https://code-boxx.com/php-user-role-management-system/ where there is a table filled with "actions" like "Edit client assigned user" (in table `permissions` from the article) and then which user role can do which action is defined in the table `roles_permissions`. 
The authorization "rules" would be defined in the database rather than in the service classes code like the first example. 

I am now asking, with your experience, what would you recommend for larger scaled applications?

**Comment by odan** [comment]
I think there is no general solution that is always right for everything. It is essentially a matter of what your specific application requires. Of course, it is also important that this solution also scales. In practice, however, you only notice this when the application actually becomes larger. Then it is sometimes even necessary to rewrite these parts. There is never a guarantee for the future. 

You can protect your application from changes to the implementation by, for example, checking the permissions using an **interface**. Then you can change the implementation behind it much easier later. This way, you can postpone the decision whether to store the roles and permissions in the database or not. I would move the permission checking to a special class that implements this interface. The interface could be called `AuthorizationCheckerInterface` and the implementation could be called `AuthorizationChecker`.

Example:

```php
<?php

namespace App\Authorization;

interface AuthorizationCheckerInterface
{
    public function isGranted(string $attribute, mixed $subject = null): bool;
}
```

```php
<?php

final class AuthorizationChecker implements AuthorizationCheckerInterface
{
    private AuthorizationCheckerRepository $repository;

    public function __construct(SessionInterface $session, AuthorizationCheckerRepository $repository)
    {
        $this->session = $session;
        $this->repository = $repository;
    }

    public function isGranted(string $attribute, mixed $subject = null): bool
    {
        $loggedInUser = $this->session->get('user');

        if (!$loggedInUser) {
            return false;
        }

        // decide, check role etc.
       $userRole = $this->userRepository->getUserRoleById($loggedInUser->id);
       
       // I would recommend implementing them with a role hierarchy. 
       // ...
        
        return true;
    }
}
```

**Usage**

```php
if (!$authorizationChecker->isGranted(UserRole::ADMIN)) {
    throw new AccessDeniedException();
}
```

**Comment by samuelgfeller** [comment]
Thanks for the answer. Okay, this would be an extended version of the first option I mentioned, and I see it making most sense for small to medium apps.   

The thing is that I already postponed the decision whether to store the roles and permissions in the database or not until now, and the time has come to decide, and it has to be suitable for big projects with a lot of components. 

Is a solution with interfaces like you presented scalable and even recommended by you for large projects, or would you choose to store the permissions in the database? 
How are you doing it in the larger projects that you have created and maintain?

Also, maybe a follow-up question, would you do it similarly as the article describes it or a bit differently?

**Comment by odan** [comment]
I would store at least the users and the use role in the database. Storing and managing the user permissions brings an extra layer of complexity that is not needed in every case. I cannot give you a general recommendation here, because it depends on so many factors. I maintain a very big application that works and scales very well with a role based (RBAC) approach. If your users need to configure the permissions, then everything should be stored in the database, of course. But again, it depends on your application. If you have further questions, you can consult me because this topic is too complex.

**Comment by samuelgfeller** [comment]
Okay, so there is not only one way even for big projects, that is an answer already. 

> If your users need to configure the permissions, then everything should be stored in the database

This is a good point. If there is no general recommendation, I should probably have asked what are the advantages vs disadvantages of both methods or when to use one versus the other so that I can make a decision based on that and not pure speculation of a novice. If you happen to think about a few points, I'd still love to hear them.   
Otherwise, I will try out a bunch of things and decide for myself what I prefer using and what is more easily testable.

**Comment by samuelgfeller** [comment]
So I concluded that unless I sat down for hours trying to think about everything and writing a very solid concept so that the permissions get updated consistently correctly, it can only be a mess once the table grows big. I'd probably be better off using a library with inbuilt tools like calculating permission objects, which is something that I'm trying to avoid where I can.

I have quite complex cases (that would scare me to maintain on a database level, and I feel I'd have to do compromises as well). I'm not entirely sure on how to implement those cases following your input on the `AuthorizationCheckerInterface`.

One frequent case is when a user with an inferior role is not allowed to change all columns from a table. E.g. user with role `advisor` is allowed to change the `status_id` from a `client` but not his `name` (or only when he's the owner by being linked with `user_id`). Role `advisor` can create a new client and set a `name` at this moment.

So there is a distinction between creating, updating, deleting and also for potentially each column of an object.
First I thought about a `ClientAuthorizationChecker` that implements `isGranted` but this is only one function, I cannot do the distinction inside there. I would need to have a `ClientUpdateAuthorizationChecker` where I pass the column that I want to change and inside the `isGranted` an if-statement to check if the authenticated role is allowed to perform the action on the given column. The owner id would also be passed as argument, or perhaps it'd even make most sense to directly pass the whole `DataObject` from the database that contains this info.  
There would be a `ClientUpdateAuthorizationChecker`, `ClientDeleteAuthorizationChecker`, `ClientCreateAuthorizationChecker`, `ClientReadAuthorizationChecker` (there are some notes linked to a client that some user roles may not see) and possibly `ClientListAuthorizationChecker` that each implement the `isGranted` method. 

Am I seeing things right, or is that not how you meant it? 
I guess another option would be to have `isGrantedToCreate`, `isGrantedToUpdate`, `isGrantedToDelete`, `isGrantedToRead` and `isGrantedToList` in the interface but then every checker class that implements it would have to have those functions declared which sometimes are not needed (when the permission rules are more simple).

**Comment by odan** [comment]
The `$attribute` parameter is basically a string. You could also change the attribute into `CREATE_PRODUCT`, `POST_SHOW` or whatever fit your domain model. Your `AuthorizationChecker` then just needs to check if the current user has this "role". So you just need one class for the security checks.

**Comment by samuelgfeller** [comment]
I'm not sure to understand. Where and how would the permission logic be (where you commented `// I would recommend implementing them with a role hierarchy.`)?

If there is only one `AuthorizationChecker` the individual permissions would have to be stored elsewhere unless the `isGranted` has a long list of if-statements for each case, but that doesn't make sense. I understood that this would be all in the code, is that not how you meant it?

I think I'd understand it if on the database side I implement a `permission` table with entries like "Create user with role user", "Change user role to advisor", "Change user role to admin", "Create client", "Change client status" etc. and a `user_role_to_permission` table where the roles that are authorized are linked to the actions in the `permission` table. Like [the article](https://code-boxx.com/php-user-role-management-system/) describes. That way, I just have to get the user role of the authenticated user and check if he's allowed to perform said action by checking the database table `user_role_to_permission`.
No more than one `isGranted` method is needed for this. 

But that would basically be the option 2 I initially talked about, and I'd prefer not having to think about a full concept on how to formulate and store these permissions in the database and maintain them. Especially the more complex things where a role is allowed to partially mutate a table. 

This following example is a real scenario, authorization checks purely in the code (like "option 1"). It seems to me easier to maintain and more simple to implement than having permissions in the database, that's why I tend to this route for now. 

But it doesn't match what you said about one `AuthorizationChecker` and the `$attribute` param, so I must have not understood you. 
I hope it can be clarified if you tell me what you think about this example and how you would implement these same rules:
```php
/**
 * Logic to check if granted to update client
 *
 * @param array $clientDataToUpdate validated array with as key the column to
 * update and value the new value. There may be one or multiple entries,
 * depending on what the user wants to update
 *
 * @param int $ownerId user_id linked to client
 *
 * @return bool
 */
public function isGrantedToUpdateClient(array $clientDataToUpdate, int $ownerId): bool
{
    $grantedUpdateKeys = [];
    if (($loggedInUserId = $this->session->get('user_id')) !== null) {
        $authenticatedUserRoleData = $this->userRoleFinderRepository->getUserRoleDataFromUser($loggedInUserId);
        /** @var array{role_name: int} $userRoleHierarchies role name as key and hierarchy value
         * (lower hierarchy number means higher privilege) */
        $userRoleHierarchies = $this->userRoleFinderRepository->getUserRolesHierarchies();
        // Roles: newcomer < advisor < managing_advisor < administrator
        // If logged-in hierarchy value is smaller or equal advisor -> granted
        if ($authenticatedUserRoleData->hierarchy <= $userRoleHierarchies['advisor']) {
            // Things that advisor is allowed to change for all client records even when not owner
            if (array_key_exists('first_name', $clientDataToUpdate)) {
                $grantedUpdateKeys[] = 'first_name';
            }
            if (array_key_exists('last_name', $clientDataToUpdate)) {
                $grantedUpdateKeys[] = 'last_name';
            }
            // Everything that owner and managing_advisor is permitted to do
            // advisor may only edit client_status_id if he's owner | managing_advisor and higher is allowed
            if ($loggedInUserId === $ownerId || $authenticatedUserRoleData->hierarchy <= $userRoleHierarchies['managing_advisor']) {
                // Check if client_status_id is among data to be changed if yes add it to $grantedUpdateKeys array
                if (array_key_exists('client_status_id', $clientDataToUpdate)) {
                    $grantedUpdateKeys[] = 'client_status_id';
                }
            }
            // Things that only managing_advisor and higher privileged are allowed to do
            if ($authenticatedUserRoleData->hierarchy <= $userRoleHierarchies['managing_advisor']) {
                if (array_key_exists('user_id', $clientDataToUpdate)) {
                    $grantedUpdateKeys[] = 'user_id';
                }
            }
        }
    }
    // If data that the user wanted to update and the grantedUpdateKeys are equal by having the same keys -> granted
    foreach ($clientDataToUpdate as $key => $value) {
        // If at least one array key doesn't exist in $grantedUpdateKeys it means that user is not permitted 
        if (!in_array($key, $grantedUpdateKeys, true)) {
            return false;
        }
    }
    return true;
}
```

**Comment by odan** [comment]
When the user is logging in, you give that users a one or more roles by loading them into the user session. Your AuthorizationChecker then reads that list of roles and checks if the user has this specific role. For even more detailed permission checks, you may also pass the column name as the optional `$subject` parameter. So the shown concept should be flexible enough to match even your requirements, because you can pass any string that represents a specific role or permission.

**Comment by samuelgfeller** [comment]
> When the user is logging in, you give that users a one or more roles by loading them into the user session

What do you mean by giving that user one or more roles by loading them into the session? I give the user one role that I store in `user_role_id` of the `user` table when I create it, or change the role when I update the user. When the user is logging in, I just retrieve the infos about his role, not giving it one. 

> Your AuthorizationChecker then reads that list of roles and checks if the user has this specific role

What are the kinds of roles you are talking about here? If the authenticated user has role "admin" for instance, how does the `AuthorizationChecker` know that `admin` is allowed to perform x and y action, even if I pass the column name in the `$subject` parameter? Where and how is this logic that I wrote as code in my comment above? 

> I would store at least the users and the use role in the database. Storing and managing the user permissions brings an extra layer of complexity that is not needed in every case.

Based on this sentence, I would think that it's not the database, but where is the permission logic else? I'm talking about the rule that x role can perform y action. Basically what the code block of my previous comment does.   
I can't imagine having the entire logic to create / update / delete of all the different components of my app in one single `AuthorizationChecker`. Only the client update logic is already more than 40 lines of code, where would that be, and what do you think about the format in which I wrote it, is that efficient or do you have a better way?

**Comment by odan** [comment]
Sorry, I think we get not further from here. Again, this topic is too specific without knowing all your project specific requirements.

**Comment by samuelgfeller** [comment]
Alright, I respect that. 

I am kind of surprised nonetheless. I don't think there is anything project specific in this whole thread. I demonstrated a typical use case with the code in [this comment ](https://github.com/odan/slim4-tutorial/issues/56#issuecomment-1283782853) and was asking how you would implement that following your input on a `AuthorizationChecker` without having the permissions stored in the database. Especially clarification on what you wrote because Marc and I don't understand how you mean it. 

I was looking for a pretty general handling of roles and permissions that would be useable for multiple different projects, and if there would be any information that you were missing, I'd have gladly provided it.

But I respect that you don't want to invest more time in this. As this is a very general question and every developer that makes an app that has users with roles will be confronted to this, I think that an addition in one of your ebooks or blog posts (with examples that illustrate what you mean) would be very valuable. And if monetization is the only issue, I'd also be ready to pay you. Had intended to do so anyway for future real project specific consulting.
